### PR TITLE
Home Owner Groups (Part 1)

### PR DESCRIPTION
**Pull request by dimak1**
*Issue #:* /bcgov/entity#13009

*Description of changes:*
- Refactor to use a composable to manage the state of the groups, add/edit owners, etc.
- Add / Remove Owners to / from Groups
- Delete Groups
- Tests are coming in the next commits

### Screenshots

Screen Shot 2022-08-08 at 3 13 36 PM

Screen Shot 2022-08-08 at 3 13 42 PM

Screen Shot 2022-08-08 at 3 13 51 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the PPR license (Apache 2.0).

### CODE REVIEW

File: ppr-ui/src/components/mhrRegistration/HomeOwners/TableGroupHeader.vue [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        cancelText: 'Cancel'
+      }"
+    />
+    <div v-if="!isEditingGroupMode" class="group-header-summary">
+      <div>
+        <span class="mr-2 font-weight-bold">Group {{ groupId }}</span>
+        |
+        <span class="ma-2">Owners: {{ owners.length }} </span>
+        |
+        <span class="ma-2">Group Tenancy Type: </span>
```
Comment (dimak1) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/TableGroupHeader.vue:24]: This will be added later as the Tenancy Type requirement gets finalized.

**Review by ChasBoucher — APPROVED** [review]
Looks good

**Review by doug-lovett — APPROVED** [review]
Very nice.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue [hunk 2]
```diff
@@ -228,13 +233,17 @@ import { focusOnFirstError } from '@/utils'
 import { MhrRegistrationHomeOwnersIF } from '@/interfaces/mhr-registration-interfaces'
 import { SimpleHelpToggle } from '@/components/common'
 /* eslint-enable no-unused-vars */
+import HomeOwnerGroups from '@/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue'
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue]: Might be able to simplify import path: `./HomeOwnerGroups.vue`
super minor

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue [hunk 3]
```diff
@@ -284,7 +299,11 @@ export default defineComponent({
     }
 
     const localState = reactive({
+      ownersGroupId: showGroups.value
+        ? getGroupForOwner(props.editHomeOwner?.id)?.groupId
+        : null,
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue]: Should probably be a computed value no?

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue [hunk 4]
```diff
@@ -173,8 +173,13 @@
             class="mt-2"
             hideAddressHint
           />
+          <hr class="mt-3 mb-10" />
+          <HomeOwnerGroups
+            :groupId="ownersGroupId"
+            :isAddingHomeOwner="isAddingHomeOwner"
+            @setOwnerGroupId="updateGroupIdForOwner($event)"
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue]: Pretty minor but if you only had one instance of this method, you could simplify with `ownerGroupId = $event` here

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue [hunk 5]
```diff
@@ -284,7 +299,11 @@ export default defineComponent({
     }
 
     const localState = reactive({
+      ownersGroupId: showGroups.value
+        ? getGroupForOwner(props.editHomeOwner?.id)?.groupId
+        : null,
       owner: { ...defaultHomeOwner },
+      ownerGroupId: undefined,
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue:314]: Why not `''` if it is of string type elsewhere?

File: ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  },
+  setup (props, { emit }) {
+    const { required } = useInputRules()
+    const { getGroupDropdownItems, showGroups } = useHomeOwners()
+
+    const localState = reactive({
+      ownerGroupId: props.groupId,
+      groups: computed(() => getGroupDropdownItems(props.isAddingHomeOwner)),
+      groupRules: computed(() =>
+        showGroups.value ? required('Select a group for this owner') : []
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue]: Probably a good idea to add `{ }` just for consistency, probably prettier at work again :)

File: ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      )
+    })
+
+    const setOwnerGroupId = groupId => {
+      emit('setOwnerGroupId', groupId)
+    }
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue]: While i'm not opposed to 'less is more', for consistency sake might be a good idea to add the `()` and an argument type, `string` in this case.

File: ppr-ui/src/composables/mhrRegistration/home-owner.ts [hunk 8]
```diff
@@ -1,9 +1,33 @@
...
+  MhrRegistrationHomeOwnersIF
+} from '@/interfaces'
+import '@/utils/use-composition-api'
+
+import { ref, computed, readonly, watch } from '@vue/composition-api'
+import { useActions, useGetters } from 'vuex-composition-helpers'
+import { find, remove, set, findIndex } from 'lodash'
+import { Watch } from 'vue-property-decorator'
+
+const DEFAULT_GROUP_ID = '1'
```
Comment (cameron-eyds) [inline ppr-ui/src/composables/mhrRegistration/home-owner.ts:11]: Noticed this was defined on the AddEditHomeOwner as well. Intentional?

File: ppr-ui/src/store/actions/actions-model.ts [hunk 9]
```diff
@@ -317,3 +318,10 @@ export const setMhrLocation: ActionIF = ({ commit }, { key, value }): void => {
 export const setCivicAddress: ActionIF = ({ commit }, { key, value }): void => {
   commit('mutateCivicAddress', { key, value })
 }
+
+export const setMhrRegistrationHomeOwnerGroups: ActionIF = (
+  { commit },
+  groups: MhrRegistrationHomeOwnerGroupIF[]
+): void => {
+  commit('mutateMhrHomeOwnerGroups', groups)
+}
```
Comment (cameron-eyds) [inline ppr-ui/src/store/actions/actions-model.ts:327]: Prettier doing funny things again i think. 
Is there a way to configure it?

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue (hunk 5 above)
Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue:314]: This is so we can use `OR ||` operator to set a default value easier.

File: ppr-ui/src/store/getters/state-getters.ts [hunk 10]
```diff
@@ -644,9 +645,16 @@ export const getMhrRegistrationHomeDescription = (state: StateIF): MhrRegistrati
...
+export const getMhrRegistrationHomeOwners = (
+  state: StateIF
+): MhrRegistrationHomeOwnersIF[] => {
+  const owners = []
+  state.stateModel.mhrRegistration.ownerGroups.forEach(group => {
+    group.owners.forEach(owner =>
+      owners.push({ ...owner, groupId: group.groupId })
+    )
+  })
+  return owners
```
Comment (cameron-eyds) [inline ppr-ui/src/store/getters/state-getters.ts]: Formatting here

File: ppr-ui/src/composables/mhrRegistration/home-owner.ts (hunk 8 above)
Comment (dimak1) [reply ppr-ui/src/composables/mhrRegistration/home-owner.ts:11]: There was a Default Owner ID, just a starting ID for Owners and Groups. All Owners would go to Group 1 automatically by default.

File: ppr-ui/src/views/newMhrRegistration/HomeOwners.vue [hunk 11]
```diff
@@ -121,68 +120,94 @@
...
+      const { id, ...editedOwner } = owner
+      homeOwners[owner.id] = editedOwner
+      setMhrRegistrationHomeOwners(homeOwners)
+    }
+
+    const removeHomeOwner = owner => {
+      const homeOwners = [...getMhrRegistrationHomeOwners.value]
+      homeOwners.splice(homeOwners.indexOf(owner), 1)
+      setMhrRegistrationHomeOwners(homeOwners)
+    }
```
Comment (cameron-eyds) [inline ppr-ui/src/views/newMhrRegistration/HomeOwners.vue]: Not major but for consistency, might be a good idea to give an argument type here so we know what `owner` is

**Review by cameron-eyds — COMMENTED** [review]
Some great stuff here!
Just a few comments mostly to do with formatting for consistency :)

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue (hunk 5 above)
Comment (cameron-eyds) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue:314]: 🤔 Where you used `OR` operator is was also defined as a string though, so i don't follow exactly.
`(localState.ownerGroupId as string) || '1'`

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue (hunk 2 above)
Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue]: Done.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue (hunk 3 above)
Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue]: Updated.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue (hunk 4 above)
Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue]: Updated.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue (hunk 6 above)
Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue]: Added.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue (hunk 7 above)
Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnerGroups.vue]: Added.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue (hunk 5 above)
Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/AddEditHomeOwner.vue:314]: Updated to remove type.

File: ppr-ui/src/store/actions/actions-model.ts (hunk 9 above)
Comment (dimak1) [reply ppr-ui/src/store/actions/actions-model.ts:327]: Yea, i think it goes over the max char per line. I've updated and saved it without formatting for now.

File: ppr-ui/src/store/getters/state-getters.ts (hunk 10 above)
Comment (dimak1) [reply ppr-ui/src/store/getters/state-getters.ts]: Updated.

File: ppr-ui/src/views/newMhrRegistration/HomeOwners.vue (hunk 11 above)
Comment (dimak1) [reply ppr-ui/src/views/newMhrRegistration/HomeOwners.vue]: Added types for owners.

**Review by cameron-eyds — APPROVED** [review]
Thanks for the updates! 

G2g after resolving the tests 👍

**Comment by dimak1** [comment]
@cameron-freshworks addressed your comments and updated tests. Should be good now 🤞🏻

File: ppr-ui/tests/unit/BaseDialog.spec.ts [hunk 12]
```diff
@@ -85,11 +82,21 @@ describe('Base Dialog tests', () => {
...
+    )
+    expect(
+      wrapper.findComponent(DialogContent).vm.$props.setHasContactInfo
+    ).toBe(false)
+    expect(wrapper.findComponent(DialogButtons).vm.$props.setAcceptText).toBe(
+      'default accept'
+    )
+    expect(wrapper.findComponent(DialogButtons).vm.$props.setCancelText).toBe(
+      'default cancel'
+    )
```
Comment (cameron-eyds) [inline ppr-ui/tests/unit/BaseDialog.spec.ts]: Are all the scenarios like this and below due to the prettier extension or is this reformatting intentional? 

Either way, could you turn off the auto formatter you are using and revert these changes.

File: ppr-ui/tests/unit/BasePartySummary.spec.ts [hunk 13]
```diff
@@ -99,16 +99,23 @@ describe('Secured Party list tests', () => {
...
-    expect(item1.querySelectorAll('td')[2].textContent).toContain('test@company.com')
+    expect(item1.querySelectorAll('td')[0].textContent).toContain(
+      'SECURED PARTY COMPANY LTD.'
+    )
+    expect(item1.querySelectorAll('td')[1].textContent).toContain(
+      '1234 Fort St.'
+    )
+    expect(item1.querySelectorAll('td')[2].textContent).toContain(
+      'test@company.com'
+    )
```
Comment (cameron-eyds) [inline ppr-ui/tests/unit/BasePartySummary.spec.ts]: same as above

File: ppr-ui/tests/unit/BaseDialog.spec.ts [hunk 14]
```diff
@@ -113,11 +120,21 @@ describe('Base Dialog tests', () => {
...
+      ).toEqual(options.textExtra || [])
+      expect(
+        wrapper.findComponent(DialogContent).vm.$props.setHasContactInfo
+      ).toBe(options.hasContactInfo || false)
+      expect(wrapper.findComponent(DialogButtons).vm.$props.setAcceptText).toBe(
+        options.acceptText
+      )
+      expect(wrapper.findComponent(DialogButtons).vm.$props.setCancelText).toBe(
+        options.cancelText
+      )
```
Comment (cameron-eyds) [inline ppr-ui/tests/unit/BaseDialog.spec.ts]: same as above

File: ppr-ui/tests/unit/BasePartySummary.spec.ts [hunk 15]
```diff
@@ -136,33 +143,47 @@ describe('Debtor list tests', () => {
...
-    expect(item1.querySelectorAll('td')[3].textContent).toContain('June 15, 1990')
+    expect(item1.querySelectorAll('td')[0].textContent).toContain(
+      'TEST 1 INDIVIDUAL DEBTOR'
+    )
+    expect(item1.querySelectorAll('td')[1].textContent).toContain(
+      '1234 Fort St.'
+    )
+    expect(item1.querySelectorAll('td')[3].textContent).toContain(
+      'June 15, 1990'
+    )
```
Comment (cameron-eyds) [inline ppr-ui/tests/unit/BasePartySummary.spec.ts]: same as above

**Review by cameron-eyds — COMMENTED** [review]
Sorry, i know it's a hassle but can you revert those changes done by the auto formatter and disable it if it's not configurable.

File: ppr-ui/tests/unit/BaseDialog.spec.ts (hunk 12 above)
Comment (dimak1) [reply ppr-ui/tests/unit/BaseDialog.spec.ts]: We should probably adjust prettier.config.js
`printWidth: 80`

Comment (cameron-eyds) [reply ppr-ui/tests/unit/BaseDialog.spec.ts]: I'm not sure, whatever you have to do to stop auto reformatting code like this 😆 
It's seems like it's doing more against you than for you at the moment.

Comment (dimak1) [reply ppr-ui/tests/unit/BaseDialog.spec.ts]: I think over time different formatters were used and it's kinda messed up (some lines 80, others 120, etc).
